### ISSUE TITLE
ReturnCount false positive on lambda returns

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#5341]
## Context
According to the (default) configuration lambda returns should be excluded (be them labeled or not). It might be also to do with nesting, not sure.

## Expected Behavior
No flagging.

## Observed Behavior
Flagged with ReturnCount = 3 (there are 4 returns: 2 in lambdas, 2 outside lambdas, 2 labeled).

## Steps to Reproduce
```kotlin
    @Test fun `function with mix of labeled returns should not count lambdas`() {
        val code = """
            fun transform(flag: Boolean, data: List<List<String>>): String {
                if (data.isEmpty()) return "empty"
                val intermediateResult = data.flatMap { list ->
                    val partial = list.flatMap inner@{ str ->
                        if (flag) {
                            return@inner emptyList()
                        }
                        str.toCharArray().toList()
                    }
                    return@flatMap partial.subList(0, partial.size / 2)
                }
                return intermediateResult.toCharArray().concatToString()
            }
        """.trimIndent()

        val findings = ReturnCount(
            TestConfig(
                mapOf(
                    // https://github.com/detekt/detekt/blob/v1.21.0/detekt-core/src/main/resources/default-detekt-config.yml#L626-L632
                    MAX to "2",
                    EXCLUDE_LABELED to "false",
                    EXCLUDE_RETURN_FROM_LAMBDA to "true",
                    EXCLUDE_GUARD_CLAUSES to "false",
                )
            )
        ).compileAndLint(code)
        assertThat(findings).isEmpty()
    }
```

## Your Environment

* Version of detekt used: 1.21.0

**Comment by sanggggg** [comment]
@TWiStErRob 
Can I work on this issue? I see codes and found that this issue is caused because [labeled lambda](https://kotlinlang.org/docs/returns.html#return-to-labels) is not considered.

But I have some suggestions before start working. I thought `excludeReturnFromLambda` and `excludeLabeled` configurations are ambiguous and overlapping each other.

---

The purpose of the `ReturnCount` rule is to limit returns for each function. (Now, especially named function)

Therefore **is return inside lambda or not** doesn't matter, but **is return for target named function** matter.
And IMAO **is labeled return or not** is enough for checking **is return for target named function**.

| . | in lambda | not in lambda |
| ------- | ------- | ------- |
| labeled return | not counted | (maybe impossible) |
| not labeled return | counted | counted |

> counted -> return for target named function
> not counted -> return for lambda or else...

So, IMHAO I suggest removing `excludeReturnFromLambda` and changing default value of `excludeLabeled` to true.
- `excludeReturnFromLambda` is ambiguous: whether return state "in" lambda or return state "for" lambda.
- And the current behavior of `excludeReturnFromLambda` is checking is labeled or not. this is overlapping `excludeLabeled` option and making confusion.

https://github.com/detekt/detekt/blob/090b5bcba74fe59732cc7fab92a70833f2f8445d/detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ReturnCount.kt#L97-L98

**Comment by sanggggg** [comment]
Well, there is other ways of lambda declaration [annoynimous function](https://kotlinlang.org/docs/lambdas.html#anonymous-functions) which returned by unlabeled return.
Considering this, how about removing `excludedLabeled` option too and enforcing detekt to detect return for only target named function?

**Comment by TWiStErRob** [comment]
You can work on it from my view, I'm a contributor like you, not a collaborator. I hope someone from @detekt can give feedback on your specs.

I personally don't think we need to remove anything, My feeling is that there's an `&&` vs `||` -like bug hiding not in play sight.

Although "enforcing detekt to detect returns for the only target named function" sounds like a good approach to me. It depends on what's the intention of the rule is:
 * any word that looks like a `return` is confusing and therefore needs to be counted  
   because it takes cognitive load to understand that that specific `return` is not a real return from the top `fun`.
 * count the number paths that can actually terminate the function

based on documentation it's the second one: https://detekt.dev/docs/rules/style/#returncount

---

The only thing I would like to highlight is that `labeled return - not in lambda` is quite simple, not impossible: you can always slap a function name on the return: `fun f() { return@f Unit }` and this return can be inside a lambda too:
```kotlin
fun f(list: List<String>) {
    list.forEach {
        if (it.isEmpty()) return@f
    }
}
```
so you might want to review your proposal.

**Comment by sanggggg** [comment]
Oh, I mistook you as a contributor because you posted a lot of good issues, sorry. @TWiStErRob 

I was not considering the perspective of cognitive errors (not only the number of matching returns) and the labeled return inside the function, thank you for letting me know about this.

If `count the number paths that can actually terminate the function` only matters, it doesn't seem to matter **is it inside the lambda** or **is it labeled**. Only counting actual terminating return matters (and it is a clear condition to detect).

How about deleting both config and detecting only counting actual terminating return? @detekt

**Comment by BraisGabin** [comment]
@sanggggg thanks for all the ditails. They really help.

In my opinion this rule should flag blocks of code weith too much `return`s. So it should treat a lambda as a function. I mean, any `return` inside a lambda should not count in the `function` that it is in. But if inside a lambda there are more than the maximum number of returns, that lambda should be flagged.

Same happens with functions inside other functions:

```kotlin
fun a() {
  fun b() {
    if(random) {
      return "a"
    } else if (random2) {
      return "b"
  }
}
```
--- 

>     * `excludeReturnFromLambda` is ambiguous: whether return state "in" lambda or return state "for" lambda.

Good point. This change could be done in its own PR. Deprecate `excludeReturnFromLambda` and replace it for a new one.

**Comment by BraisGabin** [comment]
@sanggggg thanks for all the ditails. They really help.

There are a lot of things here. First of all, the bug related by @TWiStErRob is clearly a `false` positive and it should be fixed.

Then there is the discussion about what to do with the `return` inside a lambda. I see different options here:
- Treat a lambda as an inner function -> This would deprecate `excludeReturnFromLambda` because the `return` inside a lambda would not count for the function but for the lambda.
- Keep the current behaviour. You can ignore the `return` inside a lambda or count it inside the function that contains it.

I think that any `return` inside a lambda is a code smell by itself so maybe a rule that flags all the `return` inside a lambda could be good. But we also have `LabeledExpression` that more less those the same (because all the `return` in lambdas are labeled).

So I vote to keep everything as it is right now and just fix the false positive. If I'm missing something I would prefer to create a new issue and talk about how to improve this rule. This way we don't mix different topics. I don't want to "full stop" this conversation, it's interesting and maybe we can improve this rule or even split it in two. But if you think that there is more things to say just move it to a new issue.

And @sanggggg if you want to fix the false positive, please go ahead :).

--- 

>     * `excludeReturnFromLambda` is ambiguous: whether return state "in" lambda or return state "for" lambda.

Good point. This change could be done in its own PR. Deprecate `excludeReturnFromLambda` and replace it for a new one.

**Comment by sanggggg** [comment]
I wrote a PR to fix the false-positive bug from the perspective of the second option.
> Keep the current behaviour. You can ignore the return inside a lambda or count it inside the function that contains it.

I'll think about the name of the configs or their meaning a bit more and raise an issue!
